modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+
+object GlobalSnapshotEventCutter {
+  type GlobalPeerIdEventMap = Map[PeerId, List[(Ordinal, GlobalSnapshotEvent)]]
+
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] =
+    new EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] {
+
+      def cut(events: GlobalPeerIdEventMap): F[GlobalPeerIdEventMap] =
+        needsCutting(events)
+          .ifM(
+            OrderedEvents(events)
+              .map(Iterator.unfold(_)(_.next))
+              .flatMap(removeEvents(events, _)),
+            events.pure[F]
+          )
+
+      def needsCutting(events: GlobalPeerIdEventMap): F[Boolean] =
+        events.toBinary.liftTo[F].map(_.length > maxBinarySizeBytes.value)
+
+      def removeEvents(startingEvents: GlobalPeerIdEventMap, iterator: Iterator[PeerIdEvent]): F[GlobalPeerIdEventMap] =
+        startingEvents.tailRecM { events =>
+          for {
+            eventToRemove <- MonadThrow[F].fromOption(iterator.nextOption(), new IllegalStateException("No GlobalSnapshotEvent events"))
+            reducedEvents = removeOneEvent(events, eventToRemove)
+            cutAgain <- needsCutting(reducedEvents)
+          } yield Either.cond(!cutAgain, reducedEvents, reducedEvents)
+        }
+
+      def removeOneEvent(events: GlobalPeerIdEventMap, eventToRemove: PeerIdEvent): GlobalPeerIdEventMap =
+        events.updatedWith(eventToRemove.id) {
+          case None => None
+          case Some(globalPeerIdEvent) =>
+            val globalSnapshotEvent = eventToRemove.toGlobalSnapshotEvent
+            val updatedEvents = globalPeerIdEvent.filterNot { case (_, gse) => gse == globalSnapshotEvent }
+            Option.when(updatedEvents.nonEmpty)(updatedEvents)
+        }
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdHashedEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdHashedEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      class OrderedEvents(sortedStateEvents: List[PeerIdStateChannelEvent], hashedEvents: List[PeerIdHashedEvent]) {
```

Comment:
Why not make it a regular `case class`? I think it is still possible to create the regular class without the processing in the apply function right? So it's not guarding the manual creation of such a not pre-processed object I think.

Reply:
Not making a case class isn't for guarding; what do we gain by making it a case class?

Reply:
I guess they are easier to handle and they tend to be used as DTOs which is what is this classes role here I think.

Reply:
It's now a `case class`